api/src/main/java/org/apache/iceberg/SortOrder.java

```diff
@@ -258,14 +257,20 @@ Builder addSortField(String transformAsString, int sourceId, SortDirection direc
     }
 
     public SortOrder build() {
-      if (orderId == 0 && fields.size() != 0) {
-        throw new IllegalArgumentException("Sort order ID 0 is reserved for unsorted order");
+      if (fields.isEmpty()) {
+        if (orderId != null && orderId != 0) {
+          throw new IllegalArgumentException("Unsorted order ID must be 0");
+        }
+        return SortOrder.unsorted();
       }
-      if (fields.size() == 0 && orderId != 0) {
-        throw new IllegalArgumentException("Unsorted order ID must be 0");
+
+      if (orderId != null && orderId == 0) {
```

Comment:
If `orderId == 0`, then it can't be null, so the null check is redundant.

Reply:
After this change, `orderId` is no longer a primitive. This check avoids an NPE.